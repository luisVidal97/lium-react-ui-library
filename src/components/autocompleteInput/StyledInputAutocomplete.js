import styled from 'styled-components';

export const StyledInputText = styled.input`
  margin: 5px;
  width: 100%;
  height: ${ props => props.shapeSize};
  font-size: ${ props => props.fontSize};

  border: 1px solid #ccc;
  background-color: ${ props => props.backgroundColor};
  pointer-events: ${ props => props.pointerEvent};
  border-radius:  ${ props => props.border};
  &:hover{
    border-color: ${props => props.borderColor};
  }

  &:focus{
    outline: none !important;
    border-color: ${props => props.borderColor};
    box-shadow: 0 0 5px ${props => props.borderColor};
  }
`;

export const StyledList = styled.ul`
  list-style: none;
  background-color: white;
  margin: 5px;
  padding: 0;
  position: absolute;
  z-index: 100;
  box-shadow: 0 .125rem .25rem rgba(0,0,0,.3) !important;
  top: 45px;
  width: 100%;
  max-height: 300px;
  overflow: auto;
`;

export const StyledInputListItem = styled.li`
  padding: 10px 5px;
  cursor: pointer;

  &:hover {
    background-color: #cccccc;
  }
`;
