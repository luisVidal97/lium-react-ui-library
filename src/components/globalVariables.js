
const forms = ['contained', 'outlined'];

const heights = {
  large: '50px',
  medium: '40px',
  small: '30px',
}

const borderRadius = {
  strong: '20px',
  normal: '10px',
  none: '0px'
}

const heightsInputText = {
  large: {
    shape: '40px',
    font: '1.2rem'
  },
  medium:  {
    shape: '32px',
    font: '1rem'
  },
  small: {
    shape: '24px',
    font: '.8rem'
  },
}


const colors = {
  primary: {
      normal: '#115293',
      hover: '#0A0A36',
  },
  secondary: {
    normal: '#DC004E',
    hover: '#9A0036',
  },
  danger: {
      normal: '#ff0000',
      hover: '#cc0000',
  },
  warning: {
    normal: '#ffea05',
    hover: '#e2d006'
  },
  success: {
    normal: '#06c920',
    hover: '#019e16'
  },
  dark: {
    normal: '#000000',
    hover: '#686767',
  },
};

const defaultStyle = {
  normal: '#000000',
  hover: '#686767',
}

export { forms, borderRadius, heights, heightsInputText, colors, defaultStyle } ;
