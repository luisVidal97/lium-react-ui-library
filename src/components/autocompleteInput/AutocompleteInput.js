import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { StyledInputText, StyledList, StyledInputListItem } from './StyledInputAutocomplete';
import { validationsInputAutocomplete } from './validationsInputAutoComplete';

/*
  This is an simple input with a extra functionality to suggest different words depending on you
  are writing. This is a controlled component that has his own state. Notice, there are a prop very
  important: ownItems. This is an array with the word that the user want to show like suggestions. If
  is undefined, the component will consume api that fetch similar words.

*/
const AutocompleteInput = (props) => {

  const [suggestions, setSuggestions] = useState([]);
  const [items, setItems] = useState([]);
  const [inpuText, setInpuText] = useState('');

  const sources = validationsInputAutocomplete(props);


  useEffect(() => {
    if(props.ownItems){
      const newList = [...props.ownItems]
      setItems(newList);
    }else {
      // https://www.datamuse.com/api/
      fetch(`https://api.datamuse.com/sug?s=${inpuText}`)
        .then( response => response.json() )
        .then( data => {
          const array = data.map(item => item.word);
          setItems(array);
        });
    }
  }, [inpuText])



  const handleOnChange = ({target}) =>{
    const value = target.value;
    let choosenSuggestions = [];
    if (value.length > 0){
      const reguex = new RegExp(`^${value}`,'i');
      choosenSuggestions = items.sort().filter(v => reguex.test(v));
    }
      setSuggestions(choosenSuggestions);
      setInpuText(value);
  }

  const renderSuggestions = ()  => {

    if (suggestions.length > 0){
      return (
        <StyledList>
        {
          suggestions.map( item =>
            <StyledInputListItem key={item} onClick={() => choosenSuggestions(item)}>{item}</StyledInputListItem>
          )
        }
        </StyledList>
      );
    }else {
      return null;
    }
  }

  const choosenSuggestions = (value) => {
    setSuggestions([]);
    setInpuText(value);
  }

  return (
    <div style={{position:'relative'}}>
      <StyledInputText
        maxLength={props.maxLength}
        placeholder={props.placeholder}
        {...sources}
        value={inpuText}
        onChange={ (e)=> {
          handleOnChange(e);
          if(props.onChange) props.onChange(e);
        }}
        onKeyPress={ (e)=> choosenSuggestions(e.target.value) }
      />

      {
        renderSuggestions()
      }

    </div>
  )
}

AutocompleteInput.propTypes = {
  border: PropTypes.string,
  size: PropTypes.string,
  disabled: PropTypes.string,
  color: PropTypes.string,
  placeholder: PropTypes.string,
  maxLength: PropTypes.string,
  disabled: PropTypes.string,
  onChange: PropTypes.func,
  theme: PropTypes.shape({
    primary: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
    secondary: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
    danger: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
    warning: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
    success: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
    dark: PropTypes.shape({
      normal: PropTypes.string,
      hover: PropTypes.string,
    }),
  }),
}


export default AutocompleteInput;
