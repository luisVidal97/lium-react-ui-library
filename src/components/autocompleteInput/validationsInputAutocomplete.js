import {colors, heightsInputText, borderRadius} from '../globalVariables'


export const validationsInputAutocomplete= ( { disabled: disabledInputText, color, size, border} ) =>{

  const shapeSize = size && heightsInputText[size]?
  heightsInputText[size].shape
  : heightsInputText['medium'].shape;

  const fontSize = size && heightsInputText[size]?
  heightsInputText[size].font
  : heightsInputText['medium'].font;

  let borderColor = color && colors[color]?
    colors[color].normal
    :
    colors['primary'].normal;

  border = border && borderRadius[border]?
    borderRadius[border]
    :
    borderRadius['normal'];

  let backgroundColor = 'transparent', pointerEvent = 'auto';
  if (disabledInputText === 'true') {
    backgroundColor = '#E0E0E0' ;
    pointerEvent = 'none'
    borderColor = 'transparent';
  }

  return {
    shapeSize,
    fontSize,
    borderColor,
    border,
    backgroundColor,
    pointerEvent,
  };
}

