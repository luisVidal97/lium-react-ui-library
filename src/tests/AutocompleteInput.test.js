import React from 'react';
import { shallow, mount } from'enzyme';

import 'jest-styled-components';
import renderer from 'react-test-renderer';

import { heightsInputText, colors, borderRadius } from '../components/globalVariables';
import AutocompleteInput from '../components/autocompleteInput/AutocompleteInput';
import {StyledInputText, StyledInputListItem} from '../components/autocompleteInput/StyledInputAutocomplete';


describe('Tests over <AutocompleteInput/>: features and behavior', () => {

  let wrapper, tree;

  test('should show <AutocompleteInput/> successfully', () => {
    wrapper = shallow(<AutocompleteInput />);
    expect( wrapper ).toMatchSnapshot();
  });


  test(` should show default style values of input invoking <AutocompleteInput /> component (without props)
  Default values = height: 32px; font-size: 1rem; border-radius: 10px; background-color: tranparent; pointer-events: auto;
  input[hover]border-color: #115293; input[focus]border-color: #115293;  input[focus]box-shadow: 0 0 5px #115293;`, () => {

    tree = renderer.create(<AutocompleteInput />).toJSON().children[0];
    expect(tree.type).toBe('input');
    expect(tree).toHaveStyleRule('height', heightsInputText.medium.shape);
    expect(tree).toHaveStyleRule('font-size', heightsInputText.medium.font);
    expect(tree).toHaveStyleRule('background-color', 'transparent');
    expect(tree).toHaveStyleRule('pointer-events', 'auto');
    expect(tree).toHaveStyleRule('border-radius', borderRadius.normal);
    expect(tree).toHaveStyleRule('border-color', colors.primary.normal, { modifier: ':hover' });
    expect(tree).toHaveStyleRule('border-color', colors.primary.normal, { modifier: ':focus' });
    expect(tree).toHaveStyleRule('box-shadow', `0 0 5px ${colors.primary.normal}`, { modifier: ':focus' });
  });

  test('should show default style values of text input invoking <AutocompleteInput /> (with props but wrong values)', () => {
    tree = renderer.create(<AutocompleteInput disabled='treu' size='sm' color='green' border='24' moove='large'/>).toJSON().children[0];
    expect(tree.type).toBe('input');
    expect(tree).toHaveStyleRule('height', heightsInputText.medium.shape);
    expect(tree).toHaveStyleRule('font-size', heightsInputText.medium.font);
    expect(tree).toHaveStyleRule('background-color', 'transparent');
    expect(tree).toHaveStyleRule('pointer-events', 'auto');
    expect(tree).toHaveStyleRule('border-radius', borderRadius.normal);
    expect(tree).toHaveStyleRule('border-color', colors.primary.normal, { modifier: ':hover' });
    expect(tree).toHaveStyleRule('border-color', colors.primary.normal, { modifier: ':focus' });
    expect(tree).toHaveStyleRule('box-shadow', `0 0 5px ${colors.primary.normal}`, { modifier: ':focus' });
  });

  test(`should show style values of text input invoking <AutocompleteInput /> with large size, dark color, strong border,
  55 maxLength and 'Hello' placeholder`, () => {
    const props = {
      size: 'large',
      color: 'dark',
      border:'strong',
      maxLength:'55',
      placeholder: 'Hello'
    }
    tree = renderer.create(<AutocompleteInput  {...props} />).toJSON().children[0];
    expect(tree.type).toBe('input');
    expect(tree.props.maxLength).toBe(props.maxLength);
    expect(tree.props.placeholder).toBe(props.placeholder);
    expect(tree).toHaveStyleRule('height', heightsInputText.large.shape);
    expect(tree).toHaveStyleRule('font-size', heightsInputText.large.font);
    expect(tree).toHaveStyleRule('background-color', 'transparent');
    expect(tree).toHaveStyleRule('pointer-events', 'auto');
    expect(tree).toHaveStyleRule('border-radius', borderRadius.strong);
    expect(tree).toHaveStyleRule('border-color', colors.dark.normal, { modifier: ':hover' });
    expect(tree).toHaveStyleRule('border-color', colors.dark.normal, { modifier: ':focus' });
    expect(tree).toHaveStyleRule('box-shadow', `0 0 5px ${colors.dark.normal}`, { modifier: ':focus' });
  });

  test('should show disabled  <AutocompleteInput /> regardless of color prop', () => {
    tree = renderer.create(<AutocompleteInput disabled='true' size='large' color='success' border='strong' />).toJSON().children[0];
    expect(tree.type).toBe('input');
    expect(tree).toHaveStyleRule('background-color', '#E0E0E0');
    expect(tree).toHaveStyleRule('pointer-events', 'none');
    expect(tree).toHaveStyleRule('height', heightsInputText.large.shape);
    expect(tree).toHaveStyleRule('font-size', heightsInputText.large.font);
    expect(tree).toHaveStyleRule('border-radius', borderRadius.strong);
    expect(tree).toHaveStyleRule('border-color', 'transparent', { modifier: ':hover' });
    expect(tree).toHaveStyleRule('border-color', 'transparent', { modifier: ':focus' });
    expect(tree).toHaveStyleRule('box-shadow', `0 0 5px transparent`, { modifier: ':focus' });
  });

  test('should change input and show items that are given through props like array', () => {
    const list = ['Luis', 'Monica', 'Maximiliam', 'Ricky', 'Mortin'];
    wrapper = mount(<AutocompleteInput ownItems={list} />);
    const input  = wrapper.find(StyledInputText);
    const event = {
      target: {
        value : 'Lu'
      }
    }
    input.simulate('change', event);
    const text  = wrapper.find(StyledInputListItem).at(0).text();
    expect(text).toBe(list[0]);
    const event2 = {
      target: {
        value : 'm'
      }
    }
    input.simulate('change', event2)
    const text2  = wrapper.find(StyledInputListItem).at(0).text();
    const text3  = wrapper.find(StyledInputListItem).at(1).text();
    const text4  = wrapper.find(StyledInputListItem).at(2).text();
    expect(text2).toBe(list[2]);
    expect(text3).toBe(list[1]);
    expect(text4).toBe(list[4]);
  });


})
