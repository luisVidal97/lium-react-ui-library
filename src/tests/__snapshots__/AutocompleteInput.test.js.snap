// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <AutocompleteInput/>: features and behavior should show <AutocompleteInput/> successfully 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <styled.input
    backgroundColor="transparent"
    border="10px"
    borderColor="#115293"
    fontSize="1rem"
    onChange={[Function]}
    onKeyPress={[Function]}
    pointerEvent="auto"
    shapeSize="32px"
    value=""
  />
</div>
`;

exports[`Tests over <Button/>: features and behavior should show <AutocompleteInput/> successfully 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <styled.input
    backgroundColor="transparent"
    border="10px"
    borderColor="#115293"
    fontSize="1rem"
    onChange={[Function]}
    onKeyPress={[Function]}
    pointerEvent="auto"
    shapeSize="32px"
    value=""
  />
</div>
`;
